README.dev.md

```diff
@@ -79,10 +79,68 @@ During the development cycle, three main supporting branches are used:
+   - if everything goes well, this PR will automatically be closed after the draft release is created.
+1. Navigate to [Draft Github Release](https://github.com/DeepRank/deeprank2/actions/workflows/release_github.yml)
+   on the [Actions](https://github.com/DeepRank/deeprank2/actions) tab.
+2. On the right hand side, you can select the level increase ("patch", "minor", or "major") and which branch to release from.
+   - [Follow semantic versioning conventions](https://semver.org/) to chose the level increase:
+     - `patch`: when backward compatible bug fixes were made
+     - `minor`: when functionality was added in a backward compatible manner
+     - `major`: when API-incompatible changes have been made
+   - Note that you cannot release from `main` (the default shown) using the automated workflow. To release from `main`
+     directly, you must [create the release manually](#manually-create-a-release).
+3. Visit [Actions](https://github.com/DeepRank/deeprank2/actions) tab to check whether everything went as expected.
+   - NOTE: there are two separate jobs in the workflow: "draft_release" and "tidy_workspace". The first creates the draft release on github, while the second merges changes into `dev` and closes the PR.
+     - If "draft_release" fails, then there are likely merge conflicts with `main` that need to be resolved first. No release draft is created and the "tidy_workspace" job does not run.
+     - If "draft_release" is succesfull but "tidy_workspace" fails, then there are likely merge conflicts with `dev` that are not conflicts with `main`. In the latter case, the draft release is created and it is up to the user to decide whether to proceed with finalizing the release or first resolving the conflicts with `dev` (this should never happen if the release branch was `dev`, as the only change will be the version bump).
+     - If both jobs succeed, then the draft release is created and the changes are merged into `dev` without any problems and the associated PR is closed. Also, if the release branch is different from `dev`, then that branch will be deleted from the remote repository.
+4. Navigate to the [Releases](https://github.com/DeepRank/deeprank2/releases) tab and click on the newest draft
+   release that was just generated.
+5. Click on the edit (pencil) icon on the right side of the draft release.
+6. Check/adapt the release notes and make sure that everything is as expected.
+7. Check that "Set as the latest release is checked".
+8. Click green "Publish Release" button to convert the draft to a published release on GitHub.
+   - This will automatically trigger [another GitHub workflow](https://github.com/DeepRank/deeprank2/actions/workflows/release.yml) that will take care of publishing the package on PyPi.
+
+#### Updating the token:
+
+NOTE: the current token (associated to @DaniBodor) allowing to bypass branch protection will expire on 9 July 2025. To update the token do the following:
```

Comment:
Is there no way to set up a non-expiring one? What if we use secrets.GITHUB_TOKEN for this scope as well? 

Also, the previous `actions/checkout@v3` didn't know any token. Maybe we can just stick to that instead of `actions/checkout@v4`?

Reply:
> Is there no way to set up a non-expiring one? 

There is, but [GitHub recommends against doing this](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/token-expiration-and-revocation#token-revoked-after-reaching-its-expiration-date) for security reasons (although they don't really elaborate on it). Do you want to set a non-expiring one anyway?

> What if we use secrets.GITHUB_TOKEN for this scope as well?

The reason we need a special token is because we need to bypass the branch protection of `main` and `dev`. I believe that GITHUB_TOKEN does not have admin rights to do so. I think it's possible to set up a GitHub bot for this, but I didn't manage to do that quickly and think that this is good enough for the time being.

> Also, the previous `actions/checkout@v3` didn't know any token. Maybe we can just stick to that instead of `actions/checkout@v4`?

:+1:

Reply:
> Is there no way to set up a non-expiring one? What if we use secrets.GITHUB_TOKEN for this scope as well?
> 
> Also, the previous `actions/checkout@v3` didn't know any token. Maybe we can just stick to that instead of `actions/checkout@v4`?

The reason we need a special token is because we need to bypass the branch protection of `main` and `dev`. I believe that GITHUB_TOKEN does not have admin rights to do so. I think it's possible to set up a GitHub bot for this, but I didn't manage to do that quickly and think that this is good enough for the time being.

It's possible to set non-expiring tokens, but [GitHub recommends against doing this](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/token-expiration-and-revocation#token-revoked-after-reaching-its-expiration-date) for security reasons (although they don't really elaborate on it). Do you want to set a non-expiring one anyway?

Reply:
> It's possible to set non-expiring tokens, but [GitHub recommends against doing this](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/token-expiration-and-revocation#token-revoked-after-reaching-its-expiration-date) for security reasons (although they don't really elaborate on it). Do you want to set a non-expiring one anyway?

Yes, I don't think is convenient to have to update it unless there is a very valid (and clear) reason for doing so.

Reply:
So, I am still uncomfortable setting up a permanent token (under my account). There are several security risks with using a token without an expiration date (and GitHub "strongly recommends" using an expiration). Especially if we at some point stop maintaining this repo, I would like the token to eventually be transferred to whoever is maintaining it, which could easily drop off my radar if a permanent token exists.
If you still feel strongly about having a permanent token, then could you please set one up under your account. You can follow [these instructions](https://github.com/DeepRank/deeprank2/blob/170_release_workflow_dbodor/README.dev.md#updating-the-token) and it only takes a minute or so to do.

Reply:
This is what ChatGPT says why it's good to set an expiration date:

> Creating a non-expiring personal access token (PAT) is less secure for several reasons:
> 
> 1. **Increased Risk of Compromise**: A non-expiring token that is exposed or leaked (e.g., through accidental sharing, a compromised machine, or a code repository) will remain valid indefinitely. This provides an attacker with perpetual access to your resources, increasing the damage they can cause over time.
> 
> 2. **Lack of Rotation**: Security best practices encourage regularly rotating access credentials to reduce the impact of a potential compromise. Non-expiring tokens do not enforce rotation, leading to a situation where the same token is used for an extended period, increasing the risk window.
> 
> 3. **Forgotten Tokens**: Non-expiring tokens can easily be forgotten about, especially if they're not actively monitored. Over time, these forgotten tokens can accumulate, creating a broader attack surface that is hard to manage and secure.
> 
> 4. **No Expiry-Driven Audit**: Expiring tokens encourage regular audits of which tokens are in use, who is using them, and whether they are still necessary. Without expirations, there is less incentive to review or revoke unnecessary tokens, potentially leading to unused tokens that are still valid and vulnerable.
> 
> 5. **Reduced Control Over Access**: Expiration dates allow for automatic revocation of access, helping maintain tighter control over who can access your resources and when. Non-expiring tokens eliminate this automatic control, making it easier for unauthorized access to persist.
> 
> 6. **Compliance Requirements**: Many security frameworks and compliance standards require that access credentials have expiration dates to minimize risks associated with prolonged exposure. Non-expiring tokens may lead to compliance issues, especially in regulated industries.
> 
> By setting expiration dates on personal access tokens, you mitigate these risks, ensure better security hygiene, and maintain tighter control over access to your resources.

Reply:
This is what ChatGPT says if I ask what is the trade off in such a situation:

> While security best practices advocate for expiring tokens due to the reasons you listed, there are situations—like yours—where convenience, project lifecycle, and the diminishing frequency of updates may justify using a non-expiring Personal Access Token (PAT). Here's a balanced view on whether expiration is truly necessary in your scenario and the associated risks:
>
> When Expiration is Less Critical:
> Low Sensitivity of the Repository: If the open-source project you're working on doesn't contain sensitive data, proprietary code, or anything of significant value (financial or reputational), then the impact of a compromised token is lower. In such cases, setting up a non-expiring token may be a reasonable decision if it saves you time and effort.
>
> Project Lifecycle: Since you mentioned that you're gradually stepping away from the project, the diminishing involvement in the project reduces the likelihood that you'll need to actively manage or monitor the token. In this scenario, revisiting and rotating tokens regularly may seem overburdening, especially as team members disengage over time.
>
> Limited Scope of the Token: If you scope the token very narrowly to limit its access (e.g., only giving it access to read from a single repository), the risks are reduced. An attacker with access to such a token wouldn’t be able to cause significant damage. Keeping it tightly scoped mitigates some of the broader risks typically associated with non-expiring tokens.

The best solution here would be to refresh the token without tying it to a personal account by using GitHub App. It seems relatively straightforward, it consists in [creating a GitHub App](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/registering-a-github-app), generate there a private key, save it in GitHub Secrets, and updating that via a workflow periodically. Of course I am not sure whether then this is _really_ simple as it seems. I'd say either we go for the latter solution, or I use my personal account for setting a non-expiring token. @DaniBodor

Reply:
We need to decide about this one

Reply:
We decided to add a meaningful note about this, and leave an expiring token.